src/Xamarin.Android.Build.Tasks/Utilities/MemoryStreamPool.cs

```diff
@@ -0,0 +1,63 @@
+using System.IO;
+using System.Text;
+
+namespace Xamarin.Android.Tasks
+{
+	/// <summary>
+	/// A class for pooling and reusing MemoryStream objects.
+	/// 
+	/// Based on:
+	/// https://docs.microsoft.com/dotnet/standard/collections/thread-safe/how-to-create-an-object-pool
+	/// https://docs.microsoft.com/dotnet/api/system.buffers.arraypool-1
+	/// </summary>
+	class MemoryStreamPool : ObjectPool<MemoryStream>
+	{
+		/// <summary>
+		/// Static instance across the entire process. Use this most of the time.
+		/// </summary>
+		public static readonly MemoryStreamPool Shared = new MemoryStreamPool ();
+
+		public MemoryStreamPool () : base (() => new MemoryStream ()) { }
+
+		public override void Return (MemoryStream stream)
+		{
+			// We want to throw here before base.Return() if it was disposed
+			stream.SetLength (0);
+			base.Return (stream);
+		}
+
+		/// <summary>
+		/// Creates a StreamWriter that uses the underlying MemoryStreamPool. Calling Dispose() will Return() the MemoryStream.
+		/// </summary>
+		public StreamWriter CreateStreamWriter () => CreateStreamWriter (Encoding.Default);
```

Comment:
Would it make more sense to use `MonoAndroidHelper.UTF8withoutBOM`?  *We* only ever use `Encoding.Default` *once*, in a *unit test*:

```
% git grep Encoding.Default src/Xamarin.Android.Build.Tasks 
src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/Utilities/FilesTests.cs:              Stream NewStream (string contents) => new MemoryStream (Encoding.Default.GetBytes (contents));
```

vs. use of `new UTF8Encoding`, excluding unit tests:

```
% git grep 'new.*UTF8Encoding' src/Xamarin.Android.Build.Tasks | grep -v Tests/
src/Xamarin.Android.Build.Tasks/Tasks/Aot.cs:			using (var sw = new StreamWriter (responseFile, append: false, encoding: new UTF8Encoding (encoderShouldEmitUTF8Identifier: false))) {
src/Xamarin.Android.Build.Tasks/Tasks/ClassParse.cs:						encoding: new UTF8Encoding (encoderShouldEmitUTF8Identifier: false))) {
src/Xamarin.Android.Build.Tasks/Tasks/CompileToDalvik.cs:					encoding: new UTF8Encoding (encoderShouldEmitUTF8Identifier: false))) {
src/Xamarin.Android.Build.Tasks/Tasks/GenerateLibraryResources.cs:		static readonly Encoding Encoding = new UTF8Encoding (encoderShouldEmitUTF8Identifier: false);
src/Xamarin.Android.Build.Tasks/Tasks/GeneratePackageManagerJava.cs:				var utf8Encoding = new UTF8Encoding (false);
src/Xamarin.Android.Build.Tasks/Tasks/JavaCompileToolTask.cs:						encoding:new UTF8Encoding (encoderShouldEmitUTF8Identifier:false))) {
src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocument.cs:			using (var file = new StreamWriter (filename, append: false, encoding: new UTF8Encoding (false)))
src/Xamarin.Android.Build.Tasks/Utilities/ManifestDocument.cs:			using (var file = new StreamWriter (stream, new UTF8Encoding (false), bufferSize: 1024, leaveOpen: true))
src/Xamarin.Android.Build.Tasks/Utilities/TypeMapGenerator.cs:			outputEncoding = new UTF8Encoding (false);
```

Reply:
Yeah, I think the default could use `Utf8withoutBOM`. Most of the time we generate java-related files.